castle/client.py

```diff
@@ -7,39 +7,71 @@
 from castle.commands.track import CommandsTrack
 from castle.exceptions import InternalServerError
 from castle.failover_response import FailoverResponse
+from castle.utils import timestamp
 
 
 class Client(object):
-    def __init__(self, request, options):
-        self.options = options or dict()
-        self.do_not_track = self.default_tracking()
-        self.cookies = self.setup_cookies(request)
-        self.context = self.setup_context(request)
+
+    @classmethod
+    def from_request(cls, request, options={}):
+        return cls(
+            cls.to_context(request, options),
+            cls.to_options(options)
+        )
+
+    @staticmethod
+    def to_context(request, options={}):
+        default_context = ContextDefault(request, options.get('cookies')).call()
+        return ContextMerger.call(default_context, options.get('context', {}))
+
+    @staticmethod
+    def to_options(options={}):
+        options.setdefault('timestamp', timestamp())
+        return options
+
+    @staticmethod
+    def failover_response_or_raise(options, exception):
+        if configuration.failover_strategy == 'throw':
+            raise exception
+        return FailoverResponse(options.get('user_id'), None, exception.__class__.__name__).call()
+
+    def __init__(self, context, options={}):
+        self.do_not_track = options.get('do_not_track', False)
+        self.timestamp = options['timestamp']
+        self.context = context
         self.api = Api()
 
+    def _add_timestamp_if_necessary(self, options):
+        if self.timestamp:
+            options.setdefault(self.timestamp)
+
     def authenticate(self, options):
         if self.tracked():
+            self._add_timestamp_if_necessary(options)
+            command = CommandsAuthenticate(self.context).build(options)
             try:
-                response = self.api.call(CommandsAuthenticate(self.context).build(options))
+                response = self.api.call(command)
```

Comment:
No need to create `command` variable, just pass it directly.

Reply:
@nijikon - I wanted to remove that from try block